numpy/ma/core.py

```diff
@@ -5079,24 +5007,66 @@ def mean(self, axis=None, dtype=None, out=None):
         1.5
 
         """
+        kwargs = {}
+        if keepdims is not np._NoValue:
+            kwargs['keepdims'] = keepdims
+
         if self._mask is nomask:
-            result = super(MaskedArray, self).mean(axis=axis, dtype=dtype)
-        else:
-            dsum = self.sum(axis=axis, dtype=dtype)
-            cnt = self.count(axis=axis)
-            if cnt.shape == () and (cnt == 0):
-                result = masked
+            ret = self._data.mean(axis, dtype, out, **kwargs)
+
+            to_mask = ~np.isfinite(ret)
+            if self.size == 0:
+                to_mask = True
+
+            if out is not None:
+                if isinstance(out, MaskedArray):
+                    out.__setmask__(to_mask if np.any(to_mask) else False)
+                return out
+
+            if isinstance(ret, ndarray):
+                ret = ret.view(type(self))
+                ret.__setmask__(to_mask if np.any(to_mask) else False)
+                return ret
+            else:
+                # otherwise it's a scalar
+                return ret if self.size else masked
+
+        # code largely copied from _mean in numpy/core/_methods.py
+
+        rcount = self.count(axis=axis, **kwargs)
+
+        # Cast bool, unsigned int, and int to float64 by default
+        if dtype is None and issubclass(self.dtype.type,
+                                        (np.integer, np.bool_)):
+            dtype = np.dtype('f8')
+
+        ret = self.sum(axis=axis, dtype=dtype, **kwargs)
+        if ret is masked:
+            return _return_float_masked(out)
+        with np.errstate(divide='ignore', invalid='ignore'):
+            if isinstance(ret, np.ndarray):
+                ret = np.true_divide(
+                        ret, rcount, out=ret, casting='unsafe', subok=False)
```

Comment:
It's deja vu all over again! I think we discussed `subok=False` before :)

As to why not simply do `ret /= count`, I'm not sure, and the change doesn't break any tests. But this line is copied from `_mean` in `numpy/core/_methods.py`. For now I would rather leave this alone, and just follow the leader.